File: py/p3-longest-substring-without-repeating-characters.py
import collections


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        longest = 0
        x = 0
        y = 0
        while y != len(s):
            if s[y] in s[x:y]:
                x += 1
            else:
                y += 1
                if y - x > longest:
                    longest = y - x

        return longest


class Solution4:
    def lengthOfLongestSubstring(self, s: str) -> int:
        longest = 0
        chars = collections.deque()
        for c in s:
            while c in chars:
                chars.popleft()
            chars.append(c)
            if longest < len(chars):
                longest = len(chars)
        return longest


class Solution3:
    def lengthOfLongestSubstring(self, s: str) -> int:

        if s == '':
            return 0

        longest = 1
        chars = {s[0]}
        x = 0
        y = 1
        while y != len(s):

            if s[y] in chars:
                chars.remove(s[x])
                x += 1
            else:
                chars.add(s[y])
                y += 1
                if y - x > longest:
                    longest = y - x

        return longest


class Solution2:
    def lengthOfLongestSubstring(self, s: str) -> int:
        '''First cut.
        
        Did not submit, just thought of a faster way.'''
        longest = 1
        for x in range(len(s)):
            chars = {s[x]}
            y = x + 1
            while y != len(s):
                if s[y] in chars:
                    break
                chars.add(s[y])
                y += 1
            if y - x > longest:
                longest = y - x
        return longest


if __name__ == '__main__':

    s = Solution()
    tests = (
        ("", 0),
        ("a", 1),
        ("abcabcbb", 3),
        ("bbbbb", 1),
        ("pwwkew", 3),
    )

    for test in tests:

        substr = test[0]
        exp = test[1]
        ans = s.lengthOfLongestSubstring(substr)
        if exp != ans:
            print(f'FAIL. s: {substr}, exp: {exp}, ans: {ans}')

File: cpp/p3-longest-substring-without-repeating-characters.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s)
    {
        if (s.empty()) {
            return 0;
        }
        int longest = 1;
        auto c1 = s.begin();
        auto c2 = c1 + 1;
        while (c2 != s.end()) {
            if (find(c1, c2, *c2) == c2) {
                ++c2;
                longest = std::max(longest, (int)std::distance(c1, c2));
            } else {
                ++c1;
            }
        }
        return longest;
    }
};

class Solution2 {
public:
    int lengthOfLongestSubstring(string s)
    {
        int longest = 0;
        deque<char> chars;
        for (const char c : s) {
            while (find(chars.begin(), chars.end(), c) != chars.end()) {
                chars.pop_back();
            }
            chars.push_front(c);
            longest = std::max(longest, (int)chars.size());
        }
        return longest;
    }
};

int main(int argc, char* argv[])
{
    struct TestCase {
        string s;
        int exp;
    };
    const TestCase testCases[] = {
        { "", 0 },
        { "a", 1 },
        { "abcabcbb", 3 },
        { "bbbbb", 1 },
        { "pwwkew", 3 },
    };

    Solution s;
    for (const auto& tc : testCases) {
        const int ans = s.lengthOfLongestSubstring(tc.s);
        if (tc.exp != ans) {
            cout << "FAIL. s: " << tc.s
                 << ", exp: " << tc.exp
                 << ", ans: " << ans << "\n";
        }
    }
    return 0;
}
